mypy/meet.py

```diff
@@ -51,155 +56,246 @@ def narrow_declared_type(declared: Type, narrowed: Type) -> Type:
+        return are_typed_dicts_overlapping(left, right, ignore_promotions=ignore_promotions)
+    elif isinstance(left, TypedDictType):
+        left = left.fallback
+    elif isinstance(right, TypedDictType):
+        right = right.fallback
+
+    if is_tuple(left) and is_tuple(right):
+        return are_tuples_overlapping(left, right, ignore_promotions=ignore_promotions)
+    elif isinstance(left, TupleType):
+        left = left.fallback
+    elif isinstance(right, TupleType):
+        right = right.fallback
+
+    # Next, we handle single-variant types that cannot be inherently partially overlapping,
+    # but do require custom logic to inspect.
+    #
+    # As before, we degrade into 'Instance' whenever possible.
+
+    if isinstance(left, TypeType) and isinstance(right, TypeType):
+        # TODO: Can Callable[[...], T] and Type[T] be partially overlapping?
+        return _is_overlapping_types(left.item, right.item)
+
+    if isinstance(left, CallableType) and isinstance(right, CallableType):
+        return is_callable_compatible(left, right,
+                                      is_compat=_is_overlapping_types,
+                                      ignore_pos_arg_names=True,
+                                      allow_partial_overlap=True)
+    elif isinstance(left, CallableType):
+        left = left.fallback
+    elif isinstance(right, CallableType):
+        right = right.fallback
+
+    # Finally, we handle the case where left and right are instances.
+
+    if isinstance(left, Instance) and isinstance(right, Instance):
+        if left.type.is_protocol and is_protocol_implementation(right, left):
+            return True
+        if right.type.is_protocol and is_protocol_implementation(left, right):
+            return True
+
+        # Two unrelated types cannot be partially overlapping: they're disjoint.
+        # We don't need to handle promotions because they've already been handled
+        # by the calls to `is_subtype(...)` up above (and promotable types never
+        # have any generic arguments we need to recurse on).
+        if left.type.has_base(right.type.fullname()):
+            left = map_instance_to_supertype(left, right.type)
+        elif right.type.has_base(left.type.fullname()):
+            right = map_instance_to_supertype(right, left.type)
         else:
-            return isinstance(other, CallableType) and is_subtype(other, one)
-    if experiments.STRICT_OPTIONAL:
-        if isinstance(t, NoneTyp) != isinstance(s, NoneTyp):
-            # NoneTyp does not overlap with other non-Union types under strict Optional checking
             return False
-    # We conservatively assume that non-instance, non-union, non-TupleType and non-TypeType types
-    # can overlap any other types.
+
+        if len(left.args) == len(right.args):
+            for left_arg, right_arg in zip(left.args, right.args):
+                if _is_overlapping_types(left_arg, right_arg):
+                    return True
```

Comment:
Do we need to care about variance here? Maybe add an example why `List[Union[A, B]]` and `List[Union[B, C]]` cause a danger with overloads?